### ISSUE TITLE
Use discriminator in json-schemas (not openAPI) for better error messaging?

### ISSUE THREAD

**Issue by Jamie5** [davishmcclurg/json_schemer#144]
I know that discriminator is only part of openAPI and not base json-schema, but there is no particular reason someone couldn't include a discriminator inside their json_schema file. And if this library just happened to check for that and made better error messages, I personally at least wouldn't complain.

**Comment by ahx** [comment]
If I understand version 2.0 correctly you can set the metaschema to openapi and have json_schemer check all openapi keywords, including discriminator.

```ruby
JSONSchemer.schema(myschema_that_uses_discriminator, meta_schema: JSONSchemer.openapi31)
```

**Comment by Jamie5** [comment]
Hmm when I try that it kind of works (certainly, recognizes the discriminator) but it also does strange stuff, such as when the discriminator value is something completely unknown, it tries to load it as a ref, and also the error messages aren't that specific. At least the first issue is different from what's found in https://github.com/davishmcclurg/json_schemer/blob/6c88530c594a953a11061c6abf642e451b1ad522/test/open_api_test.rb , so I suspect that the drop-in doesn't just work as-is. (and I don't want to convert everything to openapi to be able to do `JSONSchemer.openapi(...)`, which might behave more like the linked test does)

**Comment by davishmcclurg** [comment]
> Hmm when I try that it kind of works (certainly, recognizes the discriminator) but it also does strange stuff, such as when the discriminator value is something completely unknown, it tries to load it as a ref, and also the error messages aren't that specific.

Can you provide examples (schema and data) of the strange behavior? `discriminator` support is new and it's not well defined in the openapi spec, so I had to make some behavior guesses. You may also be hitting bugs.

Also, if you could provide the current error messages you're seeing as well as messages that you think would be more helpful, that would be great.

> so I suspect that the drop-in doesn't just work as-is. (and I don't want to convert everything to openapi to be able to do `JSONSchemer.openapi(...)`, which might behave more like the linked test does)

@ahx's example should work the same way that `JSONSchemer.openapi` does. You can provide `meta_schema: JSONSchemer.openapi31` or set `"$schema": "https://spec.openapis.org/oas/3.1/dialect/base"` in your schema.

**Comment by Jamie5** [comment]
So to be clear, one can use a json-schema (without components, or an openapi version, or anything in the schema), and by passing the openapi meta schema, it should work?

As one example of puzzling behavior adapted from a test (actually using openAPI functions)

```

      openapi = {
        'openapi' => '3.1.0',
        'components' => {
          'schemas' => {
            'MyResponseType' => {
              'oneOf' => [
                { '$ref' => '#/components/schemas/Cat' },
                { '$ref' => '#/components/schemas/Dog' }
              ],
              'discriminator' => {
                'propertyName' => 'petType',
                'mapping' => {
                  'c' => '#/components/schemas/Cat',
                  'd' => 'Dog',
                  'dog' => 'Dog'
                }
              },
            },
            'Cat' => {
              'type' => 'object',
              'properties' => {
                'name' => {
                  'type' => 'string'
                }
              }
            },
            'Dog' => {
              'properties' => {
                'bark' => {
                  'type' => 'string'
                }
              }
            }
          }
        }
      }

      schemer = JSONSchemer.openapi(openapi).schema('MyResponseType')
      puts(schemer.validate({ 'petType' => 'foo' }).to_json)
```

When running this I get

```
     JSONSchemer::UnknownRef:
       json-schemer://schema/foo
```

To be clear, I know very little about how discriminators work, but if someone presents an invalid schema that has a discriminator field that is not valid, I feel like it shouldn't try to guess a ref, especially if there is an explicit mapping, since then validating anything suddenly becomes much trickier. I think the line from the spec `if the discriminator value does not match an implicit or explicit mapping, no schema can be determined and validation SHOULD fail` suggests that the consequence should be failed validation rather than an exception being thrown.

**Comment by davishmcclurg** [comment]
> So to be clear, one can use a json-schema (without components, or an openapi version, or anything in the schema), and by passing the openapi meta schema, it should work?

Yep that's right.

> To be clear, I know very little about how discriminators work, but if someone presents an invalid schema that has a discriminator field that is not valid, I feel like it shouldn't try to guess a ref, especially if there is an explicit mapping, since then validating anything suddenly becomes much trickier. I think the line from the spec `if the discriminator value does not match an implicit or explicit mapping, no schema can be determined and validation SHOULD fail` suggests that the consequence should be failed validation rather than an exception being thrown.

I think you're right. The [spec](https://spec.openapis.org/oas/v3.1.0#fixed-fields-20) says:

> In both the oneOf and anyOf use cases, all possible schemas MUST be listed explicitly.

Your example uses `oneOf`, so it shouldn't be looking for schemas other than the ones listed. The `allOf` use case is where it gets tricky because the possible schemas aren't listed in the same way:

<details>
<summary>Specification `allOf` example</summary>

```yaml
components:
  schemas:
    Pet:
      type: object
      required:
      - petType
      properties:
        petType:
          type: string
      discriminator:
        propertyName: petType
        mapping:
          dog: Dog
    Cat:
      allOf:
      - $ref: '#/components/schemas/Pet'
      - type: object
        # all other properties specific to a `Cat`
        properties:
          name:
            type: string
    Dog:
      allOf:
      - $ref: '#/components/schemas/Pet'
      - type: object
        # all other properties specific to a `Dog`
        properties:
          bark:
            type: string
    Lizard:
      allOf:
      - $ref: '#/components/schemas/Pet'
      - type: object
        # all other properties specific to a `Lizard`
        properties:
          lovesRocks:
            type: boolean
```

</details>

And they describe `mapping` as potentially being a reference:

> An object to hold mappings between payload values and schema names or references.

Which I think means ref resolution is necessary at some point. I'll have to think about the best way to handle it. At a minimum, `oneOf` and `anyOf` should only look in the provided list. For `allOf`, it might make sense to catch `UnknownRef` (and `InvalidRefResolution`) and return a validation error instead.

---

> EDIT: In this example, it is also a bit surprising to me that even if you provide an explicit mapping, then `assert(schemer.valid?(MISTY.merge('petType' => 'Cat')))` should still work - it seems like if you provide an explicit mapping, then you shouldn't listen to any implicit mapping. Though I can't tell if the spec says to still do the implicit mapping as well anyways.

It is a little unclear (from the [spec](https://spec.openapis.org/oas/v3.1.0#fixed-fields-20) again):

> Here the discriminator value of dog will map to the schema #/components/schemas/Dog, rather than the default (implicit) value of Dog.

"rather than" seems to indicate the implicit mapping no longer applies.

---

If you'd like a workaround for your original `JSONSchemer::UnknownRef` problem, I think you could use a ref resolver that always returns an error:

```ruby
JSONSchemer.openapi(openapi, :ref_resolver => proc { { 'not' => true } })
```

**Comment by davishmcclurg** [comment]
I just opened a PR that I think fixes the issues you're running into: https://github.com/davishmcclurg/json_schemer/pull/145

Do you think you could test it out?

**Comment by Jamie5** [comment]
Thanks! I can confirm that when I change a test to purposely use a wrong discriminator value, and also remove the hack in the ref_resolver, it fails with the missing schema issue. Then, when using that branch instead, it now fails with a validation error as one would hope, ``value at `<path>` does not match `discriminator` schema``. (It would be nice I feel like if the data_pointer field pointed to the field itself (the one specified by `discriminator`) instead of the whole object, but I guess there's reasons for the current behavior. But nevertheless it works.

**Comment by davishmcclurg** [comment]
Fix went out in 2.1.0.
